src/megaapi_impl.cpp

```diff
@@ -23166,96 +23166,118 @@ void MegaApiImpl::sendPendingRequests()
             {
-                MegaNode* child = nodeList->get(i);
-
-                // get folder name
-                const char* name = child->getName();
-                bool gotName = deviceName == name;
+                if(deviceNameNode->type != FOLDERNODE) { e = API_EACCESS; break; }
 
                 // get 'device-id' tag
-                const char* id = child->getCustomAttr("device-id");
-                bool gotId = deviceId == id;
+                const auto& childAttrs = deviceNameNode->attrs.map;
+                nameid attrId = AttrMap::string2nameid("dev-id"); // "device-id" would be too long
+                auto found = childAttrs.find(attrId);
 
-                // determine whether we found our node or not
-                if (!gotName && !gotId) { continue; } // ignore this child folder
+                // a folder without the attribute, or with it but with different value, will not be accessed
+                if (found == childAttrs.end() || found->second != deviceId) { e = API_EACCESS; break; }
+            }
+            else // create `DEVICE_NAME` remote dir
+            {
+                vector<NewNode> newnodes(1);
+                NewNode *newnode = &newnodes[0];
 
-                if (gotName && gotId) // got it, use this one to add another backup to it
-                {
-                    deviceNameNode = child;
-                    break;
-                }
+                // set up new node as folder node
+                newnode->type = FOLDERNODE;
+                newnode->nodehandle = 0; // why not UNDEF?
 
-                // there's an error if only one of the folder's {name, id} is correct and the other is wrong.
-                e = API_EINCOMPLETE;
-                break;
+                // generate fresh random key for this folder node
+                byte buf[FOLDERNODEKEYLENGTH];
+                client->rng.genblock(buf, FOLDERNODEKEYLENGTH);
+                newnode->nodekey.assign((char*)buf, FOLDERNODEKEYLENGTH);
+                SymmCipher key;
+                key.setkey(buf);
+
+                // generate fresh attribute object with the folder name
+                AttrMap attrs;
+                fsAccess->normalize(&deviceName);
+                attrs.map['n'] = deviceName;
+                nameid attrId = AttrMap::string2nameid("dev-id"); // "device-id" would be too long
+                attrs.map[attrId] = deviceId;
+
+                // JSON-encode object and encrypt attribute string
+                string attrstring;
+                attrs.getjson(&attrstring);
+                newnode->attrstring.reset(new string);
+                client->makeattr(&key, newnode->attrstring, attrstring.c_str());
+
+                // add the newly generated folder node
+                client->putnodes(h, move(newnodes));
+
+                // Get the new node & handle.
+                // HOW ? How can it wait until the remote folder was created?
```

Comment:
You need to chain multiple API commands. Currently it's a bit tedious: the `putnodes_result()` will check the type of MegaRequest associated to the result (`TYPE_TYPE_BACKUP_FOLDER`) and proceed with the rest of the required steps (in your case, create the backup folder).
A better approach would be to adjust the putnodes to accept a lambda completion function, that can be passed to the `putnodes()` here as a parameter. Instead of the `putnodes_result()`, the `CommandPutNodes::procresult()` would call your completion function, which would proceed with the required steps.
However, the putnodes adapted for completion functions is in a branch that will take time to merge, unfortunately, so I suggest to go the old way, which perhaps is uglier, but we will avoid conflicts with the removal-SIC branch (the one that introduces the new completion function for fetchnodes).
There are examples of many `xxx_result()` that actually don't call the `fireOnRequestFinish()`, but chain another command. You can check for reference. Anyway, we can discuss on the phone, I'll show you how to do it with an existing example, if you want to.